sds_data_manager/lambda_code/IAlirtCode/ialirt_data_query_api.py

```diff
@@ -0,0 +1,245 @@
+"""I-ALiRT Data Query lambda."""
+
+import json
+import logging
+import os
+import time
+from datetime import datetime, timedelta, timezone
+
+import boto3
+from boto3.dynamodb.conditions import Key
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+table_name = os.environ.get("DATA_TABLE")
+region = os.environ.get("AWS_DEFAULT_REGION", "us-west-2")
+dynamodb = boto3.resource("dynamodb", region_name=region)
+table = dynamodb.Table(table_name)
+
+
+def apply_time_filters(params: dict, query_kwargs: dict) -> tuple | None:
+    """Apply the filters for time.
+
+    Parameters
+    ----------
+    params : dict
+        Event parameters.
+    query_kwargs : dict
+        Query keyword arguments.
+
+    Returns
+    -------
+    query_kwargs : dict
+        The updated key expression with time filters applied.
+    start : str
+        Start time.
+    end : str
+        End time.
+    """
+    key_expr = query_kwargs["KeyConditionExpression"]
+
+    start = params.get("time_utc_start") or params.get("met_in_utc_start")
+    end = params.get("time_utc_end") or params.get("met_in_utc_end")
+
+    if start and end:
+        start_dt = validate_time(start)
+        end_dt = validate_time(end)
+        if start_dt is None or end_dt is None:
+            return None
+    elif start:
+        # Calculate end to be 1 hour later.
+        start_dt = validate_time(start)
+        if start_dt is None:
+            return None
+        end_dt = start_dt + timedelta(hours=1)
+    elif end:
+        # Calculate start to be 1 hour earlier.
+        end_dt = validate_time(end)
+        if end_dt is None:
+            return None
+        start_dt = end_dt - timedelta(hours=1)
+    else:
+        return None
+
+    start = start_dt.strftime("%Y-%m-%dT%H:%M:%S")
+    end = end_dt.strftime("%Y-%m-%dT%H:%M:%S")
+
+    key_expr &= Key("time_utc").between(start, end)
+    query_kwargs["KeyConditionExpression"] = key_expr
+
+    return query_kwargs, start, end
+
+
+def validate_time(ts: str):
+    """Validate a timestamp string in ISO 8601 format (YYYY-MM-DDTHH:MM:SS).
+
+    Parameters
+    ----------
+    ts : str
+        The timestamp string to validate.
+
+    Returns
+    -------
+    datetime | None
+        A `datetime` object if the timestamp is valid,
+        or `None` if the format is invalid.
+    """
+    try:
+        return datetime.fromisoformat(ts)
+    except Exception:
```

Comment:
There is a lot of `return None` / `if X is None` checks throughout. Would it be easier to follow by raising your own Exception here that you can then catch rather than having different return types?

```python
class BadTimeException(Exception):
    pass
```
In your main code function something like this:
```
try:
    ...
except BadTimeException:
    return _error(...)
```

Catch the smallest scope which is ValueError here.
```suggestion
    except ValueError:
        raise BadTimeException from e
```

Reply:
Nice idea. I like this better.